### ISSUE TITLE
Deprecation api page in website 

### ISSUE THREAD

**Issue by rajibmitra** [kubernetes/website#18675]
**This is a Feature Request**

We need to have one separate page where it will be mentioned which api got deprecated in which k8s version. 

**What would you like to be added**

A separate page, it will help a lot the end users , they will just come to this page and find out that the api they are looking for could be deprecated because of older k8s version. 

**Why is this needed**
Better user experience 
**Comments**

It is inspired by @liggitt 's discussion in slack and Falco implementation in latest k8s cluster.

**Comment by liggitt** [comment]
Agree this would be excellent to provide. Here are some thoughts I had around the organization of this:

Categories of deprecations:
* REST API types
* REST API fields
* Command line flags
* Configuration file versions
* Configuration file fields
* Feature gates

Structuring data to answer specific questions:
* For a given release, it should be easy to see what deprecated things exist without scrubbing release notes (not just newly deprecated things, but previously deprecated things that are living out their deprecation period)
* For a given item in the category above, it should be easy to see what release it was initially deprecated in, the releases in which it existed in a deprecated state, and eventually what release it was removed in.

Maintenance:
* The more we can automate gathering and rendering of this data, the better our chances of keeping it up to date and accurate. For example, the following script generates a table of deprecated API types from the generated openapi schema for a given release:

```sh
releases=(
    v1.13.0
    v1.14.0
    v1.15.0
    v1.16.0
    v1.17.0
)
for release in ${releases[*]}; do
  echo "## $release Deprecated REST API"
  echo "| API Version | Kind | Description |"
  echo "|-------------|------|-------------|"
  git show v1.16.0:./api/openapi-spec/swagger.json | jq -r '
    .definitions[] |
    select(."x-kubernetes-group-version-kind") |
    (.description // "" | match("(deprecated.*$)";"i")) as $deprecation |
    select($deprecation) |
    ."x-kubernetes-group-version-kind"[] |
     select(.kind | test("List$") | not) |
     "| \(.group)\(if .group == "" then " " else "/" end)\(.version) | \(.kind) | \($deprecation.captures[0].string) |"
  ' | sort  
  echo ""
  echo ""
done
```

Ideally, we would have a generation mechanism for each of the categories, which would be run as part of publishing the docs for a given release.

The script above outputs the following tables (which would make sense in a per-release view of deprecated items):

## v1.13.0 Deprecated REST API
| API Version | Kind | Description |
|-------------|------|-------------|
|  v1 | Binding | Deprecated in 1.7, please use the bindings subresource of pods instead. |
| admissionregistration.k8s.io/v1beta1 | MutatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead. |
| admissionregistration.k8s.io/v1beta1 | ValidatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead. |
| apiextensions.k8s.io/v1beta1 | CustomResourceDefinition | Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead. |
| apps/v1beta1 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1beta2/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| apps/v1beta2 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta2 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| apps/v1beta2 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta2 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| extensions/v1beta1 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1beta2/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| extensions/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| extensions/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| extensions/v1beta1 | Ingress | DEPRECATED - This group version of Ingress is deprecated by networking.k8s.io/v1beta1 Ingress. See the release notes for more information. |
| extensions/v1beta1 | NetworkPolicy | DEPRECATED 1.9 - This group version of NetworkPolicy is deprecated by networking/v1/NetworkPolicy. NetworkPolicy describes what network traffic is allowed for a set of Pods |
| extensions/v1beta1 | PodSecurityPolicy | Deprecated: use PodSecurityPolicy from policy API Group instead. |
| extensions/v1beta1 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1beta2/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| scheduling.k8s.io/v1alpha1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |
| scheduling.k8s.io/v1beta1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |


## v1.14.0 Deprecated REST API
| API Version | Kind | Description |
|-------------|------|-------------|
|  v1 | Binding | Deprecated in 1.7, please use the bindings subresource of pods instead. |
| admissionregistration.k8s.io/v1beta1 | MutatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead. |
| admissionregistration.k8s.io/v1beta1 | ValidatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead. |
| apiextensions.k8s.io/v1beta1 | CustomResourceDefinition | Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead. |
| apps/v1beta1 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1beta2/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| apps/v1beta2 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta2 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| apps/v1beta2 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta2 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| extensions/v1beta1 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1beta2/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| extensions/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| extensions/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| extensions/v1beta1 | Ingress | DEPRECATED - This group version of Ingress is deprecated by networking.k8s.io/v1beta1 Ingress. See the release notes for more information. |
| extensions/v1beta1 | NetworkPolicy | DEPRECATED 1.9 - This group version of NetworkPolicy is deprecated by networking/v1/NetworkPolicy. NetworkPolicy describes what network traffic is allowed for a set of Pods |
| extensions/v1beta1 | PodSecurityPolicy | Deprecated: use PodSecurityPolicy from policy API Group instead. |
| extensions/v1beta1 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1beta2/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| scheduling.k8s.io/v1alpha1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |
| scheduling.k8s.io/v1beta1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |


## v1.15.0 Deprecated REST API
| API Version | Kind | Description |
|-------------|------|-------------|
|  v1 | Binding | Deprecated in 1.7, please use the bindings subresource of pods instead. |
| admissionregistration.k8s.io/v1beta1 | MutatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead. |
| admissionregistration.k8s.io/v1beta1 | ValidatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead. |
| apiextensions.k8s.io/v1beta1 | CustomResourceDefinition | Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead. |
| apps/v1beta1 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1beta2/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| apps/v1beta2 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta2 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| apps/v1beta2 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta2 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| extensions/v1beta1 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1beta2/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| extensions/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| extensions/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| extensions/v1beta1 | Ingress | DEPRECATED - This group version of Ingress is deprecated by networking.k8s.io/v1beta1 Ingress. See the release notes for more information. |
| extensions/v1beta1 | NetworkPolicy | DEPRECATED 1.9 - This group version of NetworkPolicy is deprecated by networking/v1/NetworkPolicy. NetworkPolicy describes what network traffic is allowed for a set of Pods |
| extensions/v1beta1 | PodSecurityPolicy | Deprecated: use PodSecurityPolicy from policy API Group instead. |
| extensions/v1beta1 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1beta2/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| scheduling.k8s.io/v1alpha1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |
| scheduling.k8s.io/v1beta1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |


## v1.16.0 Deprecated REST API
| API Version | Kind | Description |
|-------------|------|-------------|
|  v1 | Binding | Deprecated in 1.7, please use the bindings subresource of pods instead. |
| admissionregistration.k8s.io/v1beta1 | MutatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead. |
| admissionregistration.k8s.io/v1beta1 | ValidatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead. |
| apiextensions.k8s.io/v1beta1 | CustomResourceDefinition | Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead. |
| apps/v1beta1 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1beta2/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| apps/v1beta2 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta2 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| apps/v1beta2 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta2 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| extensions/v1beta1 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1beta2/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| extensions/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| extensions/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| extensions/v1beta1 | Ingress | DEPRECATED - This group version of Ingress is deprecated by networking.k8s.io/v1beta1 Ingress. See the release notes for more information. |
| extensions/v1beta1 | NetworkPolicy | DEPRECATED 1.9 - This group version of NetworkPolicy is deprecated by networking/v1/NetworkPolicy. NetworkPolicy describes what network traffic is allowed for a set of Pods |
| extensions/v1beta1 | PodSecurityPolicy | Deprecated: use PodSecurityPolicy from policy API Group instead. |
| extensions/v1beta1 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1beta2/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| scheduling.k8s.io/v1alpha1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |
| scheduling.k8s.io/v1beta1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |


## v1.17.0 Deprecated REST API
| API Version | Kind | Description |
|-------------|------|-------------|
|  v1 | Binding | Deprecated in 1.7, please use the bindings subresource of pods instead. |
| admissionregistration.k8s.io/v1beta1 | MutatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 MutatingWebhookConfiguration instead. |
| admissionregistration.k8s.io/v1beta1 | ValidatingWebhookConfiguration | Deprecated in v1.16, planned for removal in v1.19. Use admissionregistration.k8s.io/v1 ValidatingWebhookConfiguration instead. |
| apiextensions.k8s.io/v1beta1 | CustomResourceDefinition | Deprecated in v1.16, planned for removal in v1.19. Use apiextensions.k8s.io/v1 CustomResourceDefinition instead. |
| apps/v1beta1 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1beta2/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| apps/v1beta2 | ControllerRevision | DEPRECATED - This group version of ControllerRevision is deprecated by apps/v1/ControllerRevision. See the release notes for more information. ControllerRevision implements an immutable snapshot of state data. Clients are responsible for serializing and deserializing the objects that contain their internal state. Once a ControllerRevision has been successfully created, it can not be updated. The API Server will fail validation of all requests that attempt to mutate the Data field. ControllerRevisions may, however, be deleted. Note that, due to its use by both the DaemonSet and StatefulSet controllers for update and rollback, this object is beta. However, it may be subject to name and representation changes in future releases, and clients should not depend on its stability. It is primarily for internal use by controllers. |
| apps/v1beta2 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| apps/v1beta2 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| apps/v1beta2 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| extensions/v1beta1 | DaemonSet | DEPRECATED - This group version of DaemonSet is deprecated by apps/v1beta2/DaemonSet. See the release notes for more information. DaemonSet represents the configuration of a daemon set. |
| extensions/v1beta1 | Deployment | DEPRECATED - This group version of Deployment is deprecated by apps/v1beta2/Deployment. See the release notes for more information. Deployment enables declarative updates for Pods and ReplicaSets. |
| extensions/v1beta1 | DeploymentRollback | DEPRECATED. DeploymentRollback stores the information required to rollback a deployment. |
| extensions/v1beta1 | Ingress | DEPRECATED - This group version of Ingress is deprecated by networking.k8s.io/v1beta1 Ingress. See the release notes for more information. |
| extensions/v1beta1 | NetworkPolicy | DEPRECATED 1.9 - This group version of NetworkPolicy is deprecated by networking/v1/NetworkPolicy. NetworkPolicy describes what network traffic is allowed for a set of Pods |
| extensions/v1beta1 | PodSecurityPolicy | Deprecated: use PodSecurityPolicy from policy API Group instead. |
| extensions/v1beta1 | ReplicaSet | DEPRECATED - This group version of ReplicaSet is deprecated by apps/v1beta2/ReplicaSet. See the release notes for more information. ReplicaSet ensures that a specified number of pod replicas are running at any given time. |
| scheduling.k8s.io/v1alpha1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |
| scheduling.k8s.io/v1beta1 | PriorityClass | DEPRECATED - This group version of PriorityClass is deprecated by scheduling.k8s.io/v1/PriorityClass. PriorityClass defines mapping from a priority class name to the priority integer value. The value can be any valid integer. |

**Comment by zacharysarah** [comment]
@liggitt 
> For example, the following script generates a table of deprecated API types from the generated openapi schema for a given release:

I love this! ✨ It should be fairly straightforward to include that in our generation scripts. 

@tengqm @kbhawkey  👋 What do you think?

**Comment by tengqm** [comment]
I believe this content would be very useful for users, especially developers who are building applications or services based on the k8s API surface.

**Comment by rajibmitra** [comment]
#https://github.com/falcosecurity/falco/issues/993

**Comment by kbhawkey** [comment]
Yes, it would be useful to provide a summary of deprecated APIs per version (if the information can be automated from the code and kept up to date).

The `description` field provides the most detailed information. You may want a way of filtering this information.
If an API is slated for deprecation, the information is described in all versions until the API is finally removed?  For example,
`Deprecated in v1.16, planned for removal in v1.19` appears in the information for v1.16, v1.17.
`Deprecated in 1.7, please use the bindings subresource of pods instead` appears in v1.16.
Is it difficult to isolate the APIs that start to deprecate for a given release (or is this information relevant)?

**Comment by liggitt** [comment]
> Is it difficult to isolate the APIs that start to deprecate for a given release (or is this information relevant)?

If we have a set of releases, and the APIs deprecated in each, generating the following for each release is easy:
* removed APIs
* new deprecated APIs
* previously deprecated APIs

**Comment by tengqm** [comment]
@liggitt I have spent quite some weeks on digesting the API specifications, from version to version. What I have got is a mixed feeling. It would be highly desirable to have a comprehensive list of API changes (not just deprecations), including:

- API groups, APIs, or fields newly introduced;
- API groups, APIs, or fields graduated;
- API groups, APIs, or fields deprecated, i.e. the lucky ones which successfully get a "deprecated" comment;
- API groups, APIs, or fields removed;
- APIs or fields whose description got changed;

We have to consider the "well-known" annotations and the "not-so-well-known" ones, changes to query strings, so on and so forth.

If you closely check the upstream swagger.json, you will find that there are many issues. Not all of them can be easily fixed even if merging changes to API types is straightforward (obviously not the case, :().

- we are not documenting which field for a resource is required for POST but optional for PUT;
- we validate whether a field is a non-negative field but failed to say so in the API;
- we check if a field value matches some predefined constants but never document them in the API;
- we know that the `watchXYZ` APIs are deprecated by the `listXYZ` APIs but we keep adding them;
- we are sometimes using variable names rather than the values they represent, when commenting API fields;
- we say something is to be removed in 1.x but it is still there in 1.x+3;
- ...

So... my point. We may want to fix the way we are generating the API spec. If the API spec is of low quality, there is no way we can fix it in the reference document generated. The automatic way of generating API spec may need a change. We cannot rely *solely* on it. 

In summary:

- I like the idea of summarizing API changes;
- I am not sure we are able to generate an accurate list of API deprecations *automagically*.
- there are other areas which may deserve efforts than this.

**Comment by sftim** [comment]
> Categories of deprecations:
>
> -   REST API types
> -   REST API fields
> -   Command line flags
> -   Configuration file versions
> -   Configuration file fields
> -   Feature gates

Related, we could think about where to show that FlexVolume, for example, has been superseded by CSI - and other details that aren't APIs.

**Comment by gaorong** [comment]
I'm still looking for this functionality.
Any progress on this?

@rajibmitra

**Comment by rajibmitra** [comment]
Sorry for the delay, you can expect a working prototype by the end of this month or early next month. @gaorong

**Comment by liggitt** [comment]
started https://github.com/kubernetes/website/pull/26483 as a WIP with high-level info about beta version removal and any notable changes in the GA replacements

**Comment by sftim** [comment]
How about having each Kubernetes version document the deprecations for that version, and have links to see the deprecation details for other versions? Is that better or worse?

For an example of a similar approach, see https://v1-18.docs.kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/

**Comment by sftim** [comment]
/sig release
I think

**Comment by liggitt** [comment]
> How about having each Kubernetes version document the deprecations for that version, and have links to see the deprecation details for other versions? Is that better or worse?

I like being able to search a single page for a given kind if I don't remember the exact release it stopped being served, but would be happy to hear other opinions

**Comment by tengqm** [comment]
I have a work in attempt to solve this. If you do `docker run -p 8000:8000 quay.io/tengqm/kab:latest` and then navigate to 
 `http://localhost:8000/apis/history/api/1.20/`, you will be able to see the all API changes between the 1.19 and 1.20 version. Behind the scene, the API specs have been decomposed and manually revised for easier comparison.

![image](https://user-images.githubusercontent.com/5094756/107893042-dae7c580-6f63-11eb-88b9-f5a5fa15bc30.png)

**Comment by sftim** [comment]
Implemented as https://k8s.io/docs/reference/using-api/deprecation-guide/ via PR #26483

**Comment by sftim** [comment]
@rajibmitra & @liggitt thanks both - this was a really useful improvement to make.
/close
